### DISCUSSION TITLE
[V6] TypeError: Cannot read properties of undefined (reading 'query')

### DISCUSSION THREAD

**Discussion by holoyan** [adonisjs/core#4451] [Japa] [answered]
I want to write test for my package but I'm getting error when fetching record from db

```typescript

import { test } from '@japa/runner'
import { BaseModel, column } from '@adonisjs/lucid/orm'


test.group('Some title | check type', (group) => {
  group.setup(async () => {
    await setup() // this works ok - db connection, table create
  })

  group.teardown(async () => {
    await cleanupTables() // this is ok too
  })

  test('ensure dummy assert works', async ({ assert }) => {

    class User extends BaseModel {
      @column({ isPrimary: true })
      declare id: number

      @column()
      declare username: string

      @column()
      declare email: string

      @column()
      declare password: string
    }

    const a = await User.find(1) // user with id=1 exsists in db

    assert.instanceOf(a, User)
  })
})


```

```

TypeError: Cannot read properties of undefined (reading 'query')
 
   at Object.executor tests/roles/has_roles.spec.ts:44
   42|      }
   43|  
 ❯ 44|      const a = await User.find(1)
   45|  
   46|      assert.instanceOf(a, User)

   ⁃ TestRunner.#runTest
     node_modules/@japa/core/src/test/runner.ts:311
   ⁃ TestRunner.#wrapTestInTimeout
     node_modules/@japa/core/src/test/runner.ts:384
   ⁃ TestRunner.#wrapTestInRetries
     node_modules/@japa/core/src/test/runner.ts:396
   ⁃ TestRunner.run
     node_modules/@japa/core/src/test/runner.ts:443
   ⁃ Test.exec
     node_modules/@japa/core/src/test/main.ts:466
   ⁃ GroupRunner.run
     node_modules/@japa/core/src/group/runner.ts:166


```

**Comment by tomgobich** [comment]
I have not tried writing tests that create models on the fly, so I'm just spitballing here. Models need booted, so maybe try adding a boot call just before your query?

```ts
test.group('Some title | check type', (group) => {
  group.setup(async () => {
    await setup() // this works ok - db connection, table create
  })

  group.teardown(async () => {
    await cleanupTables() // this is ok too
  })

  test('ensure dummy assert works', async ({ assert }) => {

    class User extends BaseModel {
      @column({ isPrimary: true })
      declare id: number

      @column()
      declare username: string

      @column()
      declare email: string

      @column()
      declare password: string
    }

    User.boot() // <--
    const a = await User.find(1) // user with id=1 exsists in db

    assert.instanceOf(a, User)
  })
})```

**Comment by thetutlage** [comment, marked as answer]
I would suggest looking at the Auth package tests that uses Lucid models. Especially this line https://github.com/adonisjs/auth/blob/develop/tests/helpers.ts#L102

**Reply by holoyan** [reply]
I looked into Lucid package tests before posting here, but it wasn't fully clear for me. 

This one was really helpfull hint , thanks I managed to make it work.

**Comment by holoyan** [comment]
I'm getting similar issue when using `predefined` models.

>I would suggest looking at the Auth package tests that uses Lucid models. Especially this line

@thetutlage  I've followed your suggestion and it works for the models created on the fly(thanks) but I'm getting error when using models from `src/models` directory.

I'm about to finish this [package](https://github.com/holoyan/adonisjs-permissions/tree/testing) and I'm blocked becouse of this issue.
if you look into the [test case](https://github.com/holoyan/adonisjs-permissions/blob/testing/tests/roles/has_roles.spec.ts#L24) you can see that [User](https://github.com/holoyan/adonisjs-permissions/blob/testing/test-helpers/index.ts#L224) and [Post](https://github.com/holoyan/adonisjs-permissions/blob/testing/test-helpers/index.ts#L252) models are created on the fly and I can do `create` and `get` operations, but if I use `Role` model it throws error 

```
> @holoyan/adonisjs-permissions@0.0.0 quick:test
> node --import=./tsnode.esm.js --enable-source-maps bin/test.ts


   TypeError: Cannot read properties of undefined (reading 'booted')
 

   ⁃ run
     node_modules/@japa/runner/index.ts:209

```

also tried to call this methods but still getting same error
```typescript
    Role.boot()
    Role.useAdapter(db.modelAdapter())


    // again error
    await Role.create({
      slug: 'edit',
      title: 'Edit',
    }) // TypeError: Cannot read properties of undefined (reading 'booted')

```


So basicly I want to have some kind of *environment*  where I can call package methods and then check if DB operations are correct or not, something like this

```typescript

import {Acl} from 'package'
import ModelPermission from 'package/models'


// Give a user the permission to edit
await Acl.model(user).allow('edit');

// then make direct call to db and check
const row = await ModelPermission.where(/*my consition */).first()
    assert.isTrue(row.some_column)


```


Any solution?

**Comment by parker-codes** [comment]
If anyone reaching this was getting this error after running a command, you may need to enable the framework to be loaded with the `startApp` option:

```ts
static options: CommandOptions = {
  startApp: true,
};
```

[Official Docs](https://docs.adonisjs.com/guides/ace/creating-commands#command-metadata)

**Reply by dunhamjared** [reply]
OMG, thank you so so much.

**Reply by S0lide** [reply]
Thanks a lot ! It REALLY helped me out
